processor/src/main/java/org/mapstruct/ap/internal/processor/AnnotationBasedComponentModelProcessor.java

```diff
@@ -88,34 +117,119 @@ protected void adjustDecorator(Mapper mapper) {
 
         decorator.removeConstructor();
 
-
         List<Annotation> annotations = getDelegatorReferenceAnnotations( mapper );
         List<Field> replacement = new ArrayList<Field>();
         if ( !decorator.getMethods().isEmpty() ) {
             for ( Field field : decorator.getFields() ) {
-                replacement.add( replacementMapperReference( field, annotations ) );
+                replacement.add( replacementMapperReference( field, annotations, injectionStrategy ) );
             }
         }
         decorator.setFields( replacement );
     }
 
+    private AnnotatedConstructor buildAnnotatedConstructorForMapper(Mapper mapper) {
+        List<AnnotationMapperReference> mapperReferencesForConstructor = new ArrayList<AnnotationMapperReference>();
+        for ( MapperReference mapperReference : mapper.getReferencedMappers() ) {
+            mapperReferencesForConstructor.add( (AnnotationMapperReference) mapperReference );
+        }
+
+        List<Annotation> mapperReferenceAnnotations = getMapperReferenceAnnotations();
+
+        removeReferenceAnnotations( mapperReferencesForConstructor, mapperReferenceAnnotations );
+
+        return new AnnotatedConstructor(
+            mapper.getName(),
+            mapperReferencesForConstructor,
+            mapperReferenceAnnotations,
+            additionalPublicEmptyConstructor() );
+    }
+
+    private AnnotatedConstructor buildAnnotatedConstructorForDecorator(Decorator decorator) {
+        List<AnnotationMapperReference> mapperReferencesForConstructor = new ArrayList<AnnotationMapperReference>();
+
+        for ( Field field : decorator.getFields() ) {
+            if ( field instanceof AnnotationMapperReference ) {
+                mapperReferencesForConstructor.add( (AnnotationMapperReference) field );
+            }
+        }
+
+        List<Annotation> mapperReferenceAnnotations = getMapperReferenceAnnotations();
+
+        removeReferenceAnnotations( mapperReferencesForConstructor, mapperReferenceAnnotations );
+
+        return new AnnotatedConstructor(
+            decorator.getName(),
+            mapperReferencesForConstructor,
+            mapperReferenceAnnotations,
+            additionalPublicEmptyConstructor() );
+    }
+
+    /**
+     * Removes constructor parameter annotations that are already on the constructor. For example, for CDI, the
+     * javax.inject.Inject annotation is on the constructor and must not be on the constructor parameters.
+     *
+     * @param annotationMapperReferences list of annotations that are be on the constructor parameter
```

Comment:
The description here is a bit confusing to me. At the moment, I don't have an idea for something better 😟